packages/reason-css-parser/ppx/EmitType.re

```diff
@@ -0,0 +1,269 @@
+      let params = [ptyp_constr(txt @@ Lident("list"), params)];
+      mk_branch(type_, is_constructor, params);
+    };
+
+  let create_value_parser = (type_name, value) => {
+    let terminal_op = (kind, multiplier) => {
+      let (type_, is_constructor, params) =
+        switch (kind) {
+        | Keyword(name) => (first_uppercase(name), false, [])
+        | Data_type(name) =>
+          let name = value_name_of_css(name);
+          let params = [ptyp_constr(txt @@ Lident(name), [])];
+          (first_uppercase(name), true, params);
+        | Property_type(name) =>
+          let name = property_value_name(name) |> value_name_of_css;
+          let params = [ptyp_constr(txt @@ Lident(name), [])];
+          (first_uppercase(name), true, params);
+        | _ => raise(Unsupported)
+        };
+      apply_modifier(multiplier, type_, is_constructor, params);
+    };
+
+    let combinator_op = (kind, values) => {
+      switch (kind) {
+      | Xor =>
+        List.map(
+          fun
+          | Terminal(kind, multiplier) => terminal_op(kind, multiplier)
+          | _ => raise(Unsupported),
+          values,
+        )
+      | _ => raise(Unsupported)
+      };
+    };
+
+    switch (value) {
+    | Terminal(kind, multiplier) =>
+      mk_typ(type_name) @@ [terminal_op(kind, multiplier)]
+    | Combinator(kind, values) =>
+      mk_typ(type_name) @@ combinator_op(kind, values)
+    // | Group(value, multiplier) => group_op(value, multiplier)
+    // | Function_call(name, value) => function_call(name, value)
+    | _ => abstract_type(type_name)
+    };
+  };
+};
+
+let extract_ppx_content = (exp: Parsetree.expression) => {
+  switch (exp.pexp_desc) {
+  | Pexp_extension((
+      _,
+      PStr([
+        {
+          pstr_desc:
+            Pstr_eval(
+              {
+                pexp_desc: Pexp_constant(Pconst_string(value, loc, _delim)),
+                _,
+              },
+              _attrs,
+            ),
+          _,
+        },
+      ]),
+    )) => (
+      value,
+      loc,
+    )
+  | _ => failwith("expected a ppx extension")
+  };
+};
+
+let extract_variable_name = (pat: Parsetree.pattern) => {
+  switch (pat.ppat_desc) {
+  | Ppat_var({txt, _}) => txt
+  | _ => failwith("expected variable name")
+  };
+};
+
+let gen_types = bindings => {
+  let rec inner = (bindings: list(Parsetree.value_binding), acc) => {
+    switch (bindings) {
+    | [] => acc
+    | [head, ...rest] =>
+      let name = extract_variable_name(head.pvb_pat);
+      let (payload, loc) = extract_ppx_content(head.pvb_expr);
+      let type_ =
+        switch (Css_spec_parser.value_of_string(payload)) {
+        | Some(ast) =>
+          module Loc: {let loc: Location.t;} = {
+            let loc = loc;
+          };
+          module Ast_builder = Ppxlib.Ast_builder.Make(Loc);
+          module Emit = Make(Ast_builder);
+          try(Emit.create_value_parser(name, ast)) {
+          | Unsupported => Emit.abstract_type(name)
```

Comment:
Instead of rising Unsupported, why not run Emit.abstract_type inside create_value_parser?

Reply:
Some places expect a `row_field` instead of a `type_declaration` for generating a variant type, so in more complex cases where we support some features but others not (e.g `Xor` and `And`) i thought it would be simpler and make more sense to raise and catch later returning the abstract type